default/AI/ResearchAI.py

```diff
@@ -453,28 +492,45 @@ def generate_research_orders():
     rng = random.Random()
     rng.seed(fo.getEmpire().name + fo.getGalaxySetupData().seed)
 
-    research_reqs = calculate_research_requirements(empire)
+    calculate_research_requirements(empire)
     total_rp = empire.resourceProduction(fo.resourceType.research)
 
     if total_rp <= 0: # No RP available - no research.
         return
 
+    raw_priorities = {}
     priorities = {}
+    on_path_to = {}
     for tech_name in fo.techs():
-        priority = get_priority(rng, tech_name)
+        this_tech = fo.getTech(tech_name)
+        if not this_tech:
+            continue
+        raw_priorities[tech_name] = get_priority(rng, tech_name)
+    for tech_name, priority in raw_priorities.iteritems():
         if not tech_is_complete(tech_name) and priority >= 0:
-            turn_needed = float(research_reqs[tech_name][1]) / total_rp
-            priorities[tech_name] = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
+            turn_needed = max(research_reqs[tech_name][2], math.ceil(float(research_reqs[tech_name][1]) / total_rp))
+            adjusted_priority = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
```

Comment:
If I understand your calculation of ```turn_needed``` correctly (the maximum of actual ingame time remaining to research the tech or rp_cost / rp_output), then this formula is no longer valid.

Consider a research line with many theoretical prerequisites which are cheap but take some time to research. So we have something like 30 turns at 2 RP/cost (arbitrary numbers to get the point across). In this case, the formula would give the tech an extremely low priority eventhough we can obviously afford to go for that tech and should do so if it is useful.

I am not sure how we actually want to use the information about the turns it actually takes to finish a tech but this is not the way to go (unless I misunderstood your revised calculation, see above). 

As discussed before, I think we really want to calculate ```float(priority) / rp_cost``` as a first step not ```float(priority) / turn_needed```. 

I also think that the diminishing return part should not get any worse just because it takes long to actually research a tech eventhough it is cheap. In fact, it makes probably more sense to start a tech line which needs much time but is cheap rather earlier than later.
So for this calculation, if anything, I would consider the minimum of the price-output ratio and the actual time to finish the tech to account for the fact we may already have started the tech...